### PR TITLE
Fix time plugin not to skip ticks & Avoid infinite loop when using microsecond min tick size

### PR DESCRIPTION
**Pull request by 7PH**
This PR addresses two already-reported issues when using micro epoch with the time axis

# 1. Missing ticks when using time axis

The code that resets the time of the first tick does not takes into account the seconds/millis/micro steps:
image

This PR adds the following sections to reset the micro and millis when needed:
```javascript
        if (step >= timeUnitSize.millisecond) {
            d.setMicroseconds(0);
        }
        if (step >= timeUnitSize.second) {
            d.setMilliseconds(0);
        }
```

The result after the fix:
image

This also fixes another issue mentioned here:
https://github.com/flot/flot/issues/1743

# 2. Infinite loop in dateTickGenerator

Mentioned here:
https://github.com/flot/flot/issues/1792

In some cases, the date tick generator will run in an infinite loop due to a bug in the custom micro date implementation provided by Flot time plugin. 

In short: The micro epoch is rounded to 3 decimals AFTER setting the date using the default Date.setTime prototype, which means that when the micro epoch is of form 1645527365.999999999, setTime will receive 1645527365 instead of 1645527366. This caused an infinite loop within dateTickGenerator.

This PR fixes this issue by first rounding the micro epoch then calling the Date.setTime original prototype.

### CODE REVIEW

**Comment by 7PH** [comment]
Hi @atmgrifter00 , what do you think about this pull request?
Would you imagine it is possible to merge this into a new version at some point in the future?
Let me know if I can help do something more, so that this fix is included in one of the next version bump 😊

**Comment by atmgrifter00** [comment]
> 
Could you fix the current CI issue with it? I believe the blank line above the comment `// Round epoch to 3 decimal accuracy` simply needs to be removed.

**Comment by 7PH** [comment]
Hi again @atmgrifter00 ,

Took me a bit of time to find the time to come back to this issue, but we are still very interested in this PR getting merged as the version we currently use diverge from the master branch only for this fix to be included.

Please find a bit of context for the unit tests that I just added. Both tests fail prior to this PR and succeed after this fix.

## Time axis tick resets (changes to lines 398 to 402)

```javascript
var d = [[0.95, 1], [1.95, 2]];
$.plot("#placeholder", [d], {
    xaxis: { mode: 'time' },
});
```

JSFiddle as proof of concept:
- Before fix: https://jsfiddle.net/powerhat/df2j8n0u/
- After fix: https://jsfiddle.net/powerhat/p49n5e8b/23/

An unit test has been added to `tests/jquery.flot.time.Test.js` for this

## MicroDate issue causing infinite loop in time tick generation (changes lines 34-35)

After copying `CreateMicroSecondDate` from `jquery.flot.time.js`, the issue can be reproduced with this snippet:

```javascript
// OK
const date1 = CreateMicroSecondDate(Date, 0.999); // Should be 0.999 as there is support for microseconds
console.log('no rounding                ', (0.999).toFixed(4), '->', date1.getTime().toFixed(3)); // 0.999

// OK
const date2 = CreateMicroSecondDate(Date, 0.4999); // Should be 0.500 as value should be rounded to the nearest microsecond (=0.500)
console.log('rounding, same second      ', (0.4999).toFixed(4), '->', date2.getTime().toFixed(3)); // 0.500

// KO
const date3 = CreateMicroSecondDate(Date, 0.9999); // Should be 1.000 as value should be rounded to the nearest microsecond (=1.000)
console.log('rounding, changing second  ', (0.9999).toFixed(4), '->', date3.getTime().toFixed(3)); // 0.000
```

JSFiddle as proof of concept:
- Before fix: https://jsfiddle.net/powerhat/o9b8wvy6/4/
- After fix: https://jsfiddle.net/powerhat/a3e0rpLo/2/

`CreateMicroSecondDate` is exposed through `$.plot.dateGenerator`, so I was able to add a unit test for this too in `tests/jquery.flot.time.Test.js`

_____

Hope that helps! Please let me know if I can do something to ease the completion of this pull request 😊

**Comment by 7PH** [comment]
Hi @atmgrifter00 , did you have the time to take a look at this? Let me know if there's more I can do to help

**Comment by atmgrifter00** [comment]
I'll try and look at it now. Note that the Travis CI build is failing, but I couldn't really figure out why at first glance.

File: source/jquery.flot.time.js [hunk 1]
```diff
@@ -31,11 +31,11 @@ API.txt for details.
 
         var oldSetTime = newDate.setTime.bind(newDate);
         newDate.update = function(microEpoch) {
-            oldSetTime(microEpoch);
-
             // Round epoch to 3 decimal accuracy
             microEpoch = Math.round(microEpoch * 1000) / 1000;
 
+            oldSetTime(microEpoch);
```
Comment (atmgrifter00) [inline source/jquery.flot.time.js:37]: Can you summarize the need for this particular rearrangement?

File: tests/jquery.flot.time.Test.js [hunk 2]
```diff
@@ -56,6 +56,17 @@ describe('A Flot chart with absolute time axes', function () {
...
     });
 
+    it('creates enough time ticks', function() {
+        plot = createPlotWithAbsoluteTimeAxis(placeholder, [[[0.95, 1], [1.95, 2]]], undefined, 'seconds');
+
+        var ticks = plot.getAxes().xaxis.ticks;
+        expect(ticks.length).toBeGreaterThan(10);
+        ticks.sort((a, b) => a.v - b.v);
+        for (var i = 1; i < ticks.length; i++) {
+            expect(ticks[i].v - ticks[i - 1].v).toBeLessThan(0.5);
```
Comment (atmgrifter00) [inline tests/jquery.flot.time.Test.js:66]: This seems like a strange test, at least at face value. Sorting ticks prior to doing any assertions on the ticks that depend on that order doesn't seem like a good idea (even if the particular test being written works). The question that immediately comes to mind is "Why does one need to sort the ticks before asserting on a condition respective to the values between each successive tick?".

Perhaps another way to write the test to provide the same confidence in the implementation is that you have the expected number of ticks (already present in test), and that the first tick is a particular number, and that the last tick is a particular number. The in-between tick values are less important and we can rely on other tests (hopefully) that that logic is well tested.

Am I misunderstanding a necessary aspect of what is being tested here?

**Review by atmgrifter00 — COMMENTED** [review]
Just a couple of lingering questions, and potentially an ask to change a test.

File: source/jquery.flot.time.js (hunk 1 above)
Comment (7PH) [reply source/jquery.flot.time.js:37]: Thanks for the quick review! I'll try to be as concise as possible but it was not easy to debug, so it's not easy to sum up in a few words 😁

Currently, the rounding of the microEpoch is made **after** forwarding the call to the default `Date.setTime`. And Date.setTime does not round the given value to the nearest millisecond but instead truncates it. Example: `new Date().setTime(0.9999)` returns `0`, not `1`. But in the custom micro epoch implementation, the value is rounded, not truncated. So calling 

I made a simple demo show-casing this behavior here:
current: https://jsfiddle.net/powerhat/o9b8wvy6/4/
after fix: https://jsfiddle.net/powerhat/a3e0rpLo/2/

In addition, this small issue is sometimes causing infinite loops in `dateTickGenerator`.
https://github.com/flot/flot/blob/0b84bfb148922c000768064a2391f2ba4a5d97bc/source/jquery.flot.time.js#L464
Sometimes, by mis-chance, the next tick will be `0.9999` which should be rounded to `1` (nearest micro epoch). Because this value is passed to the micro date `setTime` which has the issue described above and in the jsfiddles, the time is rolled back to `0`. 

Hope that helps

**Comment by atmgrifter00** [comment]
@7PH, I believe some older versions of some dependencies are now causing build issues. I will attempt to resolve these so we can move forward with your changes, but there could definitely be a delay.

File: tests/jquery.flot.time.Test.js (hunk 2 above)
Comment (7PH) [reply tests/jquery.flot.time.Test.js:66]: Hi again @atmgrifter00 

I'm not a fan of sorting ticks either. Reason for this is/was:

1. Sorting ticks is already made in this same file:
https://github.com/flot/flot/blob/0b84bfb148922c000768064a2391f2ba4a5d97bc/tests/jquery.flot.time.Test.js#L78

2. Avoid using "hard-coded" values in the test suite. I actually hesitated to use the approach you suggested (counting ticks & checking first/last tick values) based on test values. I agree it's testing more shallowly too. So, test would look like this with the other approach:

Test would be:
```javascript
    it('creates enough time ticks', function() {
        plot = createPlotWithAbsoluteTimeAxis(placeholder, [[[0.95, 1], [1.95, 2]]], undefined, 'seconds');

        var ticks = plot.getAxes().xaxis.ticks;
        expect(ticks.length).toEqual(14);
        expect(ticks[0].v).toEqual(0.95);
        expect(ticks[ticks.length - 1].v).toEqual(1.95);
    });
```
I'm happy to change it - Waiting for your green light to push

**Comment by 7PH** [comment]
Thank you for the review @atmgrifter00 - I will be waiting for your green light to push the update to the unit test - just let me know 

We use Flot at [CERN](https://cern.ch), so having this merged at some point in the future will greatly help us as our version of flot currently diverges from the latest version, but this is not blocking for us.

File: tests/jquery.flot.time.Test.js (hunk 2 above)
Comment (atmgrifter00) [reply tests/jquery.flot.time.Test.js:66]: Eh. Hard-coded values aren't anathema to me in tests. They provide the clearest expectation for what the test in question is meant to illustrate. Granted, when writing tests with hard-coded values it can be beneficial to provide multiple varied sets of input values.

Another important aspect of the change you are making that really isn't captured in the test, is the fact that the ticks are distributed across the length of the plot area, instead of more-or-less one half of it, which is the real behavior of interest (not the amount of ticks that were generated). It's unclear to me what the best way to test this would be. It might be sufficient to check that the min tick is less than 1/4 of the way along the plot area width, and a similar check for the max tick (for a given size of the plot area).

**Comment by atmgrifter00** [comment]
> Thank you for the review @atmgrifter00 - I will be waiting for your green light to push the update to the unit test - just let me know
> 
> We use Flot at [CERN](https://cern.ch), so having this merged at some point in the future will greatly help us as our version of flot currently diverges from the latest version, but this is not blocking for us.

Yeah, I'd go ahead and push an update to the test. You might consider the other vector of testing I mentioned above as well.

File: tests/jquery.flot.time.Test.js (hunk 2 above)
Comment (7PH) [reply tests/jquery.flot.time.Test.js:66]: Yea, to capture this, sorting the ticks and checking that there's no "bigger gap than a certain value, given the size of plot area, between each two consecutive ticks" was my original idea - it has the disadvantage of testing more than what the original bug was causing:
```javascript
        ticks.sort((a, b) => a.v - b.v);
        for (var i = 1; i < ticks.length; i++) {
            expect(ticks[i].v - ticks[i - 1].v).toBeLessThan(0.5);
        }
```

For reference flot generates these ticks for this test-case:
- Before the fix: `[0.95, 1.8, 1.9, 2, 1.95]`
- After fix: `[0.95, 0.9, 1, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2, 1.95]`

Given our discussion about sorting ticks and your suggestion, this test could be
```javascript
    it('creates enough time ticks', function() {
        plot = createPlotWithAbsoluteTimeAxis(placeholder, [[[0.95, 1], [1.95, 2]]], undefined, 'seconds');

        var ticks = plot.getAxes().xaxis.ticks;
        expect(ticks.length).toEqual(14);
        expect(ticks[0].v).toEqual(0.95);
        expect(ticks[ticks.length - 1].v).toEqual(1.95);
        var minTickValue = Math.min(...ticks.map(tick => tick.v));
        expect(minTickValue).toBeLessThan(1.45);
    });
```

1.2 being min value + 25% of plot width. I don't think it's necessary to test the max tick as it was not caused by the issue described in this PR, but I let it to your judgement if you want to add it.

Comment (atmgrifter00) [reply tests/jquery.flot.time.Test.js:66]: So, by "plot area width" I'm referring to the pixel width, not a relative width. To me this is less confusing as to what we are attempting to test. Something like this (I think):
```
const plotWidth = plot.width();
const minTickLocation = plot.p2c(ticks[0]);
expect(minTickLocation).toBeLessThan(plot.offset() + (plotWidth / 4));
etc...
```

I don't feel _too_ strongly about this, but to me it's easier to understand what's being tested. With the approach you currently have, I would recommend adding some comments as to what the expect means. Or, perhaps just changing the test name to `creates time ticks in expected locations` would help?

**Comment by atmgrifter00** [comment]
@7PH, could you sync to the master branch, and push again?

File: tests/jquery.flot.time.Test.js (hunk 2 above)
Comment (7PH) [reply tests/jquery.flot.time.Test.js]: Okay, I am pushing now the updated test. Added check for plot area width, with a few minor adjustements:
- Using `ticks[1]` rather than `ticks[0]` as `ticks[0]` is always the left edge thus correctly located, the first generated tick is `ticks[1]` and this was the one incorrect prior to this PR.
- Used es5 `var` to stay consistent with the rest of the file

In summary here is the test now:
```javascript
    it('creates time ticks in expected locations', function() {
        plot = createPlotWithAbsoluteTimeAxis(placeholder, [[[0.95, 1], [1.95, 2]]], undefined, 'seconds');

        var ticks = plot.getAxes().xaxis.ticks;
        expect(ticks.length).toEqual(14);

        // Check that the first generated tick is less than 1/4 of the way along the plot area width
        var firstTickLocation = plot.getAxes().xaxis.p2c(ticks[1].v);
        expect(firstTickLocation).toBeLessThan(plot.offset().left + (plot.width() / 4));
    });
```

Both expects fail before the PR and pass after. What do you think?

**Comment by 7PH** [comment]
> could you sync to the master branch, and push again?

@atmgrifter00 , I just did, and also pushed the updated test. Tests are passing now. Will be waiting for your feedback regarding the state of this PR

**Review by atmgrifter00 — APPROVED** [review]

**Comment by atmgrifter00** [comment]
> > could you sync to the master branch, and push again?
> 
> @atmgrifter00 , I just did, and also pushed the updated test. Tests are passing now. Will be waiting for your feedback regarding the state of this PR

Version 4.2.3 should be ready imminently. Thanks for the contribution!

**Comment by 7PH** [comment]
Thanks!
